src/shared/logger.ts

```diff
@@ -0,0 +1,82 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as vscode from 'vscode'
+import * as winston from 'winston'
+
+let logSettings: LogSettings
+
+class LogSettings {
+    private readonly outputChannel: vscode.OutputChannel
+    private readonly logger: winston.Logger
+    private readonly logPath: string
+
+    public constructor(outputChannel: vscode.OutputChannel, logPath: string, level: string = 'info') {
+        this.outputChannel = outputChannel
+        this.logger = winston.createLogger({
+            transports: [new winston.transports.File({ filename: logPath })]
+        })
+        this.logPath = logPath
+        // TODO: check if a log level exists--we can add a field for this to the settings page.
+        this.setLevel(level)
+    }
+
+    public getOutputChannel(): vscode.OutputChannel {
+        return this.outputChannel
+    }
+
+    public getLogger(): winston.Logger {
+        return this.logger
+    }
+
+    public setLevel(level: string): void {
+        // TODO: Ensure log level is valid
+        this.logger.level = level
+    }
+
+    public getLevel(): string {
+        return this.logger.level
+    }
+
+    public getLogPath(): string {
+        return this.logPath
+    }
+}
+
+export function initialize(context: vscode.ExtensionContext): void {
+    const outputChannel: vscode.OutputChannel = vscode.window.createOutputChannel('AWS Toolkit Logs')
+    const logPath = context.logPath + '.log'
+    logSettings = new LogSettings(outputChannel, logPath)
+
+    // TODO: localize this
+    logSettings.getOutputChannel().appendLine(`Error logs for this session can be found in ${logPath}`)
+    // TODO: register `goto logs for current session` command in command palette
+}
+
+// TODO: Format correctly/uniformly
+// TODO: Add timestamps
+// TODO: Add ability to handle arbitrary values, through meta parameter or LogEntry parameter
+export function log(level: string, message: string): void {
+    logSettings.getLogger().log(level, message)
+    logSettings.getOutputChannel().appendLine(`[${level}]: ${message}`)
```

Comment:
Would it make sense to write a winston transport for VS Code output channels instead of doing it manually?

Reply:
It would be awesome is winston handled formatting and text color for the OutputChannel.

Reply:
I'll take a look to see if that's possible.

Reply:
I have a rudimentary transport right now, but I'm not 100% sure if the formatted message is accessible (which seems extremely strange to me). Still looking into it though--this would be nice because then we can just return the `winston` logger object and to have handlers for all of the different levels and mechanisms.